### PR TITLE
Library for OLM operator-inspect functionality

### PR DESCRIPTION
**Pull request by shawn-hurley**
1. Make a library for multiple commands to enable the inspection of a given OLM operator
2. Design for the location, public interface.

### CODE REVIEW

File: enhancements/olm/operator-inspect.md [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+1. This enhancement defines a library, which will be used to create commands in multiple other projects. I think that initially the `operator-sdk` and `opm` tools are the correct fit, but eventually this could be apart of the `oc adm inspect` command.
```
[the remark this answers is not in the document]
Comment (shawn-hurley) [reply enhancements/olm/operator-inspect.md]: Sounds good, this is probably a two-step implementation just like catalog building. Build the library and then integrate into `oc`

File: enhancements/olm/operator-inspect.md [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+## Proposal
+
+A library, in the [api](https://github.com/operator-framework/api/tree/master/pkg) repo that will be vendored by `opm` and `operator-sdk`. This library will include all of the traversal logic for retrieving OLM Operators data. 
+
+The library will include a set of pflags that can be used by commands to create the options structure.
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Constraints
```
[the remark this answers is not in the document]
Comment (shawn-hurley) [reply enhancements/olm/operator-inspect.md:64]: > `this tool is about the first shotgun gathering so you only have to ask the customer once.

we want to gather everything. Just the CSV misses all the things that the install plan created (RBAC, Prometheus etc). 

This may be useful to implement in the library in the future, but I think it should be left off the for now.

**Comment by sferich888** [comment]
@deads2k didn't we already cover this in https://github.com/openshift/enhancements/blob/master/enhancements/oc/must-gather.md#must-gather-images where each operator or component can define an independant collection image.

**Comment by shawn-hurley** [comment]
@sferich888 I personally see this as something that is apart of the core platform and should, therefore, be apart of client-side inspect command to be able to handle subscriptions and installed operators.

**Comment by shawn-hurley** [comment]
To be more clear, I believe that this should be handled like a route: 
https://github.com/openshift/oc/blob/955d759071e99a30ff22a204c9276b77a371202e/pkg/cli/admin/inspect/resource.go#L97

When we find a subscription we do what this proposal is asking for us to do.

File: enhancements/olm/operator-must-gather.md [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+## Summary
+
+Components developing [must gather images]("../oc/must-gather.md#must-gather-images") to gather all the information that they will need for debugging has become the defacto mechanism for customers to father information for a support case. When debugging installation of an operator with OLM we need to gather data from objects across the cluster.
```
Comment (ecordell) [inline enhancements/olm/operator-must-gather.md:33]:
```suggestion
Components developing [must gather images]("../oc/must-gather.md#must-gather-images") to gather all the information that they will need for debugging has become the defacto mechanism for customers to gather information for a support case. When debugging installation of an operator with OLM we need to gather data from objects across the cluster. 
```

File: enhancements/olm/operator-must-gather.md [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+## Open Questions [optional]
+
+## Summary
+
+Components developing [must gather images]("../oc/must-gather.md#must-gather-images") to gather all the information that they will need for debugging has become the defacto mechanism for customers to father information for a support case. When debugging installation of an operator with OLM we need to gather data from objects across the cluster. 
+
+## Motivation
+
+Today back and forth between the customer and the operator team or the OLM team to get info about an operator failing install. This the current state or each operator needs to write their own must gather to capture everything. Operator authors should have a starting point for writing their own must gathers such that gathering the OLM information for their operator should be consistent.
```
Comment (ecordell) [inline enhancements/olm/operator-must-gather.md:37]:
```suggestion
Today, there is back and forth between the customer and the operator team or the OLM team to get info about an operator failing install. Operator authors should have a starting point for writing their own must gathers such that gathering the OLM information for their operator will be consistent. 
```

File: enhancements/olm/operator-must-gather.md [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+Components developing [must gather images]("../oc/must-gather.md#must-gather-images") to gather all the information that they will need for debugging has become the defacto mechanism for customers to father information for a support case. When debugging installation of an operator with OLM we need to gather data from objects across the cluster. 
+
+## Motivation
+
+Today back and forth between the customer and the operator team or the OLM team to get info about an operator failing install. This the current state or each operator needs to write their own must gather to capture everything. Operator authors should have a starting point for writing their own must gathers such that gathering the OLM information for their operator should be consistent. 
+
+
+### Goals
+
+List the specific goals of the proposal. How will we know that this has succeeded?
```
Comment (ecordell) [inline enhancements/olm/operator-must-gather.md:42]:
```suggestion

```

File: enhancements/olm/operator-must-gather.md [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+1. Gathering should gather exactly the operator, and relatedObjects for that operator, and the resources created by OLM.
+2. A library that can be used to enable this functionality across multiple commands.
+3. A small set of user facing CLI flags. 
+4. Integaration in a seperate must-gather image for other operator authors, or via the library should
```
Comment (ecordell) [inline enhancements/olm/operator-must-gather.md:46]: Should what?

File: enhancements/olm/operator-must-gather.md [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+
+### Non-Goals
+
+1. Anything related to the wider cluster. This is not meant to be a OCP debug tool.
+2. Specficics for a given operator. This is meant to help debug install of the operator, not the operator's functionality. 
+3. Handle low level install scenarios, such as manually creating a CSV after applying RBAC.
+
+## Proposal
+
+A library, in the [api](https://github.com/operator-framework/api/tree/master/pkg) repo that will be vendored by `OLM` and `operator-sdk`. This library will include all of the traversal logic for retrieving OLM Operators data.
```
Comment (ecordell) [inline enhancements/olm/operator-must-gather.md:57]: I'm not 100% sure this is the right place, but my opinion will be based on what changes would be required in `https://github.com/operator-framework/api/blob/master/go.mod`

(we want the api types to be importable in as many places as possible)

File: enhancements/olm/operator-must-gather.md [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+2. Specficics for a given operator. This is meant to help debug install of the operator, not the operator's functionality. 
+3. Handle low level install scenarios, such as manually creating a CSV after applying RBAC.
+
+## Proposal
+
+A library, in the [api](https://github.com/operator-framework/api/tree/master/pkg) repo that will be vendored by `OLM` and `operator-sdk`. This library will include all of the traversal logic for retrieving OLM Operators data. 
+
+The library will include a set of pflags that can be used by commands to create the options structure.
+
+OLM will create a must-gather image and productize it. Allowing customers to run this must gather when needed.
```
Comment (ecordell) [inline enhancements/olm/operator-must-gather.md:61]: Upstream or downstream?

Will it be part of the ocp release payload?

File: enhancements/olm/operator-must-gather.md [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+* The library will have a default version of this interface internally
+
+```
+type ObjectWriter interface {
+  // object is the object to be written. The implmenter can choose the output format.
+  // file is the relative path, from the starting directory of the inspect command, for the object.
+  Write(object interface{}, file string) error
+
+  // Used to determine if the root directory to be used by the library is valid.
+  IsDirViable(dir string) error
```
Comment (ecordell) [inline enhancements/olm/operator-must-gather.md:83]: This seems unrelated to the writer interface - I would probably make two interfaces that can be implemented by consumers.

File: enhancements/olm/operator-must-gather.md [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+
+## Design Details
+* Considering that many different CLI's will be interacting with the library, the library will create an interface that the CLI's will need to implement to write files to disk.
+* The library will have a default version of this interface internally
+
+```
+type ObjectWriter interface {
+  // object is the object to be written. The implmenter can choose the output format.
+  // file is the relative path, from the starting directory of the inspect command, for the object.
+  Write(object interface{}, file string) error
```
Comment (ecordell) [inline enhancements/olm/operator-must-gather.md:80]: Why not just have an `io.Writer` (or some appropriate variant)? is it critical that consumers specify a filesystem path?

File: enhancements/olm/operator-must-gather.md [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+
+The Inspector interface will be responsible for gather the following resource types:
+1. Subscription 
+2. InstallPlan
+3. Resources created from the install plan (CSV's, RBAC, CRD)
+4. Operator Deployment created by CSV
+
+
+### Test Plan
+- an e2e test will be used to validate that the command can:
```
Comment (ecordell) [inline enhancements/olm/operator-must-gather.md:120]: I think it will be easier to test if the fs is optional via a writer you pass in (tests can be written that just write to a buffer)

File: enhancements/olm/operator-must-gather.md [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+
+## Motivation
+
+Today back and forth between the customer and the operator team or the OLM team to get info about an operator failing install. This the current state or each operator needs to write their own must gather to capture everything. Operator authors should have a starting point for writing their own must gathers such that gathering the OLM information for their operator should be consistent. 
+
+
+### Goals
+
+List the specific goals of the proposal. How will we know that this has succeeded?
+1. Gathering should gather exactly the operator, and relatedObjects for that operator, and the resources created by OLM.
```
Comment (ecordell) [inline enhancements/olm/operator-must-gather.md:43]: It's a little unclear to me how related objects fit in.

Will this include a change to OLM so that subscription.status contains relatedObjects? Or this is saying that the olm must-gather image will collect OLM stuff by default, and users should run `oc inspect` to collect the operator-specific data from "custom" relatedObjects?

File: enhancements/olm/operator-must-gather.md (hunk 7 above)
Comment (shawn-hurley) [reply enhancements/olm/operator-must-gather.md:56]: that sounds like a plan if we end up bringing in to many dependencies then we can re-evaluate the location.

File: enhancements/olm/operator-must-gather.md (hunk 8 above)
Comment (shawn-hurley) [reply enhancements/olm/operator-must-gather.md:60]: Yes, it should be apart of the release payload, so that disconnected users would get the image and could run it.

I think we can come back to if an upstream image makes sense. Let's assume this is a downstream only image for now?

File: enhancements/olm/operator-must-gather.md (hunk 9 above)
Comment (shawn-hurley) [reply enhancements/olm/operator-must-gather.md:82]: +1 this was leftover from the inspect integration

File: enhancements/olm/operator-must-gather.md (hunk 12 above)
Comment (shawn-hurley) [reply enhancements/olm/operator-must-gather.md:42]: it was a leftover line. removing it

**Comment by ecordell** [comment]
OLM resources are now included in must-gather - but it stops with the objects themselves, not related objects (i.e. deployment, rbac) and doesn't consider the ownership graph of a specific operator (i.e. CRs, objects owned by CRs). This enhancement would still be valuable if it implemented those next steps.
